File: android/app/src/main/kotlin/dev/odeon/android/ui/locadora/CaixaEm3D.kt
package dev.odeon.android.ui.locadora

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.requiredSize
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch
import dev.odeon.android.ui.Cores

/// A caixa em três dimensões, e **o dedo é o controle**.
///
/// ## Decidido em 05/08/2026
///
/// > «no modo locadora os itens vêm em 3D? a capa em 3D, o cd, o vhs a fita
/// > etc… igual no web. usar o touch para isso vai ser muito legal.»
///
/// O que existia antes eram duas camadas em pose fixa — bonito e imóvel, porque
/// as camadas não dividiam ponto de fuga. Toda a explicação está no
/// `Projecao.kt`; aqui é só o uso dela.
///
/// ## Cada lado continua sendo um `Composable`
///
/// E é o ganho de não ter ido pra OpenGL: a capa é o mesmo `AsyncImage` do Coil,
/// com o mesmo cache e o mesmo token de mídia. O que a matriz faz é desenhar
/// esse conteúdo **no plano da face**, com a perspectiva de todos os outros.
///
/// O `drawWithContent` é o que permite isso: ele deixa aplicar uma matriz 4×4
/// arbitrária em volta do desenho do filho, coisa que a `graphicsLayer` não
/// aceita — ela expõe `rotationX`/`rotationY` soltos, que é justamente o que não
/// compõe.
@Composable
fun CaixaEm3D(
    largura: Dp,
    altura: Dp,
    espessura: Dp,
    modifier: Modifier = Modifier,
    /// A pose de fora, quando alguém quer controlá-la (o palco anima a entrada).
    /// `null` deixa a caixa cuidar da própria pose.
    poseControlada: Pose? = null,
    /// Se o dedo pode girar. Na estante não pode: lá o arrasto é da fileira, e
    /// disputar o gesto faria a lista não rolar.
    giravel: Boolean = false,
    /// Quanto a tampa está aberta, em graus. Zero é a caixa fechada.
    abertura: Float = 0f,
    aoTocar: (() -> Unit)? = null,
    /// Tocar na **metade direita** — a aresta oposta à dobradiça. É o gesto de
    /// abrir, e ele é separado do `aoTocar` porque são coisas diferentes: um
    /// pega a caixa, o outro a abre.
    aoTocarNaAbertura: (() -> Unit)? = null,
    /// O conteúdo de cada lado. Recebe o quanto aquele lado pega de luz — quem
    /// desenha decide o que fazer com isso, e todos escurecem por igual.
    lado: @Composable (Lado, Float) -> Unit,
) {
    /// ## O giro com inércia — «melhore o movimento»
    ///
    /// A primeira versão era `animateFloatAsState` sobre um estado: o dedo
    /// empurrava, e ao soltar a caixa voltava ao repouso por uma mola. Faltava a
    /// metade que faz um objeto parecer objeto — **ele continua andando quando a
    /// mão sai**.
    ///
    /// Agora são dois `Animatable`:
    ///
    /// 1. enquanto o dedo está na tela, `snapTo` — o giro segue o dedo sem
    ///    atraso, porque uma mola no meio do arrasto é a caixa chegando depois
    ///    da mão;
    /// 2. ao soltar, `animateDecay` com a velocidade do gesto — a caixa **gira
    ///    solta**, e um empurrão forte dá a volta e mostra o verso;
    /// 3. quando a inércia acaba, ela **assenta na face mais próxima** (capa ou
    ///    verso), na pose de três quartos.
    ///
    /// O passo 3 é o que impede a caixa de parar de perfil, que é o único ângulo
    /// em que ela não é nada — nem capa, nem verso, uma linha.
    val giroYAnim = remember { androidx.compose.animation.core.Animatable(Pose.POSE_DE_REPOUSO_Y) }
    val giroXAnim = remember { androidx.compose.animation.core.Animatable(Pose.POSE_DE_REPOUSO_X) }
    val escopo = androidx.compose.runtime.rememberCoroutineScope()

    val poseDesenhada = poseControlada ?: Pose(giroYAnim.value, giroXAnim.value)

    val densidade = LocalDensity.current
    val medidas = with(densidade) {
        Medidas(largura.toPx(), altura.toPx(), espessura.toPx())
    }
    val distancia = medidas.largura * 8f

    val haptico = LocalHapticFeedback.current

    Box(
        modifier = modifier
            /// A caixa ocupa **a largura mais a espessura**: a lombada nasce à
            /// esquerda da capa e precisa de chão. Sem isso ela é recortada pela
            /// borda do próprio elemento, e o defeito parece um erro de arte.
            .width(largura + espessura)
            .height(altura)
            .then(
                if (!giravel) {
                    Modifier
                } else {
                    Modifier.pointerInput(Unit) {
                        /// A velocidade do gesto, medida pelo próprio Compose —
                        /// é ela que vira inércia ao soltar.
                        val rastreador = androidx.compose.ui.input.pointer.util.VelocityTracker()

                        detectDragGestures(
                            onDragStart = { rastreador.resetTracking() },
                            onDragEnd = {
                                /// Graus por segundo, da velocidade em pixels.
                                val vx = rastreador.calculateVelocity().x *
                                    Pose.GRAU_POR_PIXEL_HORIZONTAL
                                escopo.launch {
                                    /// Pra onde o gesto apontava — ver `faceAlvo`.
                                    val destino = Pose.faceAlvo(giroYAnim.value, vx)
                                    giroYAnim.animateTo(
                                        targetValue = destino,
                                        animationSpec = spring(
                                            dampingRatio = 0.78f,
                                            stiffness = androidx.compose.animation.core.Spring.StiffnessLow,
                                        ),
                                    )
                                    /// Chegando, o ângulo volta à forma canônica.
                                    /// É invisível — mesma pose — e é o que
                                    /// impede o número de crescer sem fim em quem
                                    /// fica girando a caixa.
                                    giroYAnim.snapTo(Pose.daVolta(giroYAnim.value))
                                }
                                escopo.launch {
                                    giroXAnim.animateTo(
                                        targetValue = Pose.POSE_DE_REPOUSO_X,
                                        animationSpec = spring(),
                                    )
                                }
                            },
                            onDragCancel = {
                                escopo.launch {
                                    giroYAnim.animateTo(Pose.faceAlvo(giroYAnim.value, 0f), spring())
                                }
                                escopo.launch { giroXAnim.animateTo(Pose.POSE_DE_REPOUSO_X, spring()) }
                            },
                            onDrag = { mudanca, arrasto ->
                                mudanca.consume()
                                rastreador.addPosition(mudanca.uptimeMillis, mudanca.position)
                                escopo.launch {
                                    /// ## ⚠️ A régua do sinal: **a superfície
                                    /// segue o dedo**
                                    ///
                                    /// «Parece que a caixa vai pro lado contrário
                                    /// do dedo», disse o dono. A primeira reação
                                    /// foi inverter a horizontal — e estava
                                    /// errada: quem estava invertido era o eixo
                                    /// **vertical**, e o horizontal já obedecia.
                                    ///
                                    /// A conta que decide não é o gosto, é a
                                    /// projeção. O centro da capa está em
                                    /// `(0, 0, e/2)`, e depois de girado ele cai
                                    /// em `x = (e/2)·sen(giroY)` e
                                    /// `y = (e/2)·sen(giroX)`. Ou seja: **giro
                                    /// positivo move a capa pra direita e pra
                                    /// baixo**. Pra ela seguir o dedo, os dois
                                    /// eixos somam o arrasto — e é só isso.
                                    ///
                                    /// O vertical vinha subtraindo, com um
                                    /// comentário que dizia «invertido de
                                    /// propósito». Era invertido por engano.
                                    ///
                                    /// Sem `daVolta` aqui: durante o arrasto o
                                    /// ângulo é **contínuo**, e é o que permite ao
                                    /// encaixe saber pra que lado a mão ia.
                                    giroYAnim.snapTo(
                                        giroYAnim.value + arrasto.x * Pose.GRAU_POR_PIXEL_HORIZONTAL,
                                    )
                                    giroXAnim.snapTo(
                                        (giroXAnim.value + arrasto.y * Pose.GRAU_POR_PIXEL_VERTICAL)
                                            .coerceIn(-Pose.TETO, Pose.TETO),
                                    )
                                }
                            },
                        )
                    }
                },
            )
            .then(
                if (aoTocar == null) {
                    Modifier
                } else {
                    Modifier.pointerInput(Unit) {
                        detectTapGestures {
                            /// O tique seco de mexer num objeto — o mesmo da
                            /// versão anterior, e continua sendo o leve: pegar a
                            /// fita é que leva a batida.
                            haptico.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                            aoTocar()
                        }
                    }
                },
            )
            .then(
                if (aoTocarNaAbertura == null) {
                    Modifier
                } else {
                    Modifier.pointerInput(Unit) {
                        detectTapGestures { onde ->
                            /// A metade direita é a abertura; a esquerda é a
                            /// dobradiça. Tocar na dobradiça não faz nada de
                            /// propósito — um gesto que funciona nos dois lados
                            /// não ensina de que lado a caixa abre.
                            if (onde.x > size.width / 2f) aoTocarNaAbertura()
                        }
                    }
                },
            ),
    ) {
        /// Do mais fundo pro mais próximo — a ordem do pintor. Ver
        /// `ladosVisiveis`.
        /// Do mais fundo pro mais próximo — a ordem do pintor. Ver
        /// `ladosVisiveis`.
        ladosVisiveis(poseDesenhada, medidas, abertura).forEach { qual ->
            val (larguraDoLado, alturaDoLado) = tamanhoDoLado(qual, medidas)
            val luz = luzNoLado(qual, poseDesenhada, medidas, abertura)

            /// A homografia deste lado: os quatro cantos do retângulo do
            /// conteúdo, mapeados nos quatro cantos projetados.
            ///
            /// ⚠️ `setPolyToPoly` e não uma `Matrix` do Compose, e a primeira
            /// tentativa provou por quê: a conversão do Compose pra
            /// `android.graphics.Matrix` recusa componente em `z`, e a
            /// prateleira apareceu **vazia**. Ver `Projecao.kt`.
            val homografia = remember(qual, poseDesenhada, medidas, abertura) {
                val origem = floatArrayOf(
                    0f, 0f,
                    larguraDoLado, 0f,
                    larguraDoLado, alturaDoLado,
                    0f, alturaDoLado,
                )
                val destino = cantosNaTela(
                    lado = qual,
                    pose = poseDesenhada,
                    m = medidas,
                    distancia = distancia,
                    abertura = abertura,
                    centroX = larguraDoLado / 2f,
                    centroY = alturaDoLado / 2f,
                )
                android.graphics.Matrix().apply {
                    setPolyToPoly(origem, 0, destino, 0, 4)
                }
            }

            Box(
                modifier = Modifier
                    .align(Alignment.Center)
                    .requiredSize(
                        with(densidade) { larguraDoLado.toDp() },
                        with(densidade) { alturaDoLado.toDp() },
                    )
                    .drawWithContent {
                        drawIntoCanvas { it.nativeCanvas.save() }
                        drawIntoCanvas { it.nativeCanvas.concat(homografia) }
                        drawContent()
                        drawIntoCanvas { it.nativeCanvas.restore() }
                    },
            ) {
                lado(qual, luz)
            }
        }
    }
}

/// A sombra que a luz da face joga por cima do conteúdo.
///
/// Um véu preto com a opacidade do que **falta** de luz. Aplicado por cima de
/// qualquer conteúdo — pôster, papel da contracapa, madeira da lombada —, é o
/// que faz as quatro faces parecerem o mesmo objeto sob a mesma lâmpada, em vez
/// de quatro desenhos separados.
@Composable
fun BoxScope.VeuDeLuz(luz: Float) {
    if (luz >= 0.999f) return
    Box(
        Modifier
            .fillMaxSize()
            .background(Color.Black.copy(alpha = (1f - luz).coerceIn(0f, 0.6f))),
    )
}

/// A moldura padrão de um lado: fundo escuro e o véu por cima.
@Composable
fun LadoSimples(
    luz: Float,
    cor: Color = Cores.fundoElevado,
    conteudo: @Composable BoxScope.() -> Unit = {},
) {
    Box(Modifier.fillMaxSize().background(cor)) {
        conteudo()
        VeuDeLuz(luz)
    }
}

File: ios/Odeon/Interface/CaixaEm3D.swift
import SwiftUI

/// A caixa em três dimensões, e **o dedo é o controle**.
///
/// ## De onde ela vem
///
/// > «no modo locadora os itens vêm em 3D? a capa em 3D, o cd, o vhs a fita
/// > etc… igual no web. usar o touch para isso vai ser muito legal.»
///
/// O que eu tinha entregue eram duas faces em pose quase fixa — exatamente o que
/// o `Projecao.kt` do Android descreve como o estado **anterior** ao conserto de
/// lá. Copiei o defeito junto com o comentário.
///
/// ## Cada lado continua sendo uma `View`
///
/// É o ganho de não ter ido pra Metal: a capa é o mesmo `AsyncImage`, com o mesmo
/// cache e o mesmo token de mídia. O que a homografia faz é desenhar esse
/// conteúdo **no plano da face**, com a perspectiva de todas as outras.
///
/// ⚠️ `projectionEffect` aceita a matriz pronta e **não** impõe câmera própria —
/// é por isso que ele serve e o `rotation3DEffect` não.
struct CaixaEm3D<Face: View>: View {
    let medidas: Medidas
    /// Se o dedo pode girar. ⚠️ **Na estante não pode**: lá o arrasto é da
    /// fileira, e disputar o gesto faria a lista não rolar. Foi decisão do
    /// Android e vale igual aqui.
    var giravel: Bool = false
    var abertura: CGFloat = 0
    /// O interior é de **disco** (cubo no fundo) ou de **fita** (berço do
    /// cassete)? ⚠️ Só importa com a caixa **aberta** — fechada, interior não
    /// existe. É o único pedaço de «arte» que a geometria carrega, porque o forro
    /// é desenhado aqui e não nas faces.
    var interiorDeDisco: Bool = true
    /// A cor do casco, pro lábio da tampa: plástico preto no disco, papelão
    /// tingido na fita. Quem sabe essa cor é quem desenha as faces.
    var ehEscuro: Bool = true
    /// A pose de fora, quando alguém quer controlá-la. `nil` deixa a caixa cuidar
    /// da própria.
    var poseControlada: Pose?
    /// Recebe o lado e **quanta luz ele pega** — a luz escurece por igual, e é o
    /// que faz a lombada ser a mesma matéria da capa, vista de canto.
    @ViewBuilder let face: (Lado, CGFloat) -> Face

    @State private var pose = Pose()
    @State private var poseAoPegar: Pose?
    /// A inércia: ao soltar, a caixa **continua girando** e para onde o atrito a
    /// deixar. Sem isso ela para na hora, e objeto que para na hora é desenho.
    @State private var inercia: Task<Void, Never>?

    /// A câmera está a **oito larguras**. Mais perto exagera a perspectiva e a
    /// caixa vira um cone; mais longe achata e o 3D some.
    private var distancia: CGFloat { medidas.largura * 8 }

    /// O quadro precisa caber a caixa **girada**, e a diagonal é o pior caso.
    private var quadro: CGSize {
        let d = sqrt(medidas.largura * medidas.largura + medidas.espessura * medidas.espessura)
        return CGSize(width: d * 1.15, height: medidas.altura * 1.08)
    }

    private var poseAtual: Pose { poseControlada ?? pose }

    var body: some View {
        let centro = CGPoint(x: quadro.width / 2, y: quadro.height / 2)

        ZStack(alignment: .topLeading) {
            /// ## O forro — o interior que a tampa aberta revela
            ///
            /// ⚠️ **Sem ele, abrir a caixa mostra um buraco.** As faces só são
            /// desenhadas de frente (o recorte de costas da projeção), o que é
            /// exato numa caixa **fechada**: ela é sólida, o interior não existe.
            /// Com a tampa a 118° o olho passa a ver por dentro, e o que há lá é o
            /// fundo do palco — a caixa aberta vira uma lombada flutuando no nada.
            /// É o «cenário de teatro visto de trás» voltando pela porta da tampa.
            ///
            /// O forro são as faces **de costas** desenhadas como quadriláteros
            /// chapados: sem homografia, sem conteúdo, só o polígono na cor do
            /// plástico interno. É o que um estojo aberto mostra de verdade — o
            /// lado de dentro não é impresso.
            ///
            /// ⚠️ Fechada, nada disto roda: o custo do forro só existe enquanto há
            /// interior pra ver.
            if abertura > 0.5 { forro(centro: centro) }

            /// ⚠️ **Da mais funda pra mais próxima** — a ordem do pintor. Sem ela a
            /// lombada some por baixo da capa em vez de encostar nela.
            ForEach(Array(ladosVisiveis(pose: poseAtual, m: medidas, abertura: abertura).enumerated()),
                    id: \.offset) { _, lado in
                let tamanho = tamanhoDoLado(lado, medidas)
                let cantos = cantosNaTela(lado, pose: poseAtual, m: medidas,
                                          distancia: distancia, abertura: abertura, centro: centro)
                if let matriz = Homografia.mapeando(largura: tamanho.width,
                                                    altura: tamanho.height, para: cantos) {
                    face(lado, luzDoLado(lado, pose: poseAtual, m: medidas, abertura: abertura))
                        .frame(width: tamanho.width, height: tamanho.height)
                        /// ## ⚠️ **Recortar antes de projetar**, e não é zelo
                        ///
                        /// A homografia mapeia **o que é desenhado**, não o quadro.
                        /// Conteúdo que transborda o `frame` é mapeado junto, e o
                        /// resultado é uma face do tamanho errado — foi o que o
                        /// título deitado da lombada fez: reportou 292pt de largura
                        /// numa face de 24, e a lombada saiu quinze vezes maior.
                        ///
                        /// O `clipped` é o que garante que a face projetada é a
                        /// face, e não o que por acaso coube dentro dela.
                        .clipped()
                        .projectionEffect(ProjectionTransform(matriz))
                }
            }
        }
        .frame(width: quadro.width, height: quadro.height)
        /// ⚠️ A sombra é **caída pra baixo**, não em volta: caixa pousada na
        /// prateleira, não flutuando. Objeto que flutua vira cartão.
        .shadow(color: .black.opacity(0.6), radius: 9, x: 3, y: 9)
        .contentShape(.rect)
        .gesture(giravel ? arrasto : nil)
    }

    /// O forro e o lábio da concha.
    private func forro(centro: CGPoint) -> some View {
        Canvas { contexto, _ in
            let deCostas = Lado.allCases
                .filter { !deFrente($0, pose: poseAtual, m: medidas, abertura: abertura) }
                .sorted { profundidade($0, pose: poseAtual, m: medidas, abertura: abertura)
                    < profundidade($1, pose: poseAtual, m: medidas, abertura: abertura) }

            for lado in deCostas {
                let pontos = cantosNaTela(lado, pose: poseAtual, m: medidas,
                                          distancia: distancia, abertura: abertura,
                                          centro: centro)
                /// ⚠️ Três tons, e eles são **a luz do palco**: a tampa por dentro
                /// é o forro mais claro — está virada pra cima, de cara pro facho,
                /// e é isso que faz a tampa aberta ser vista como tampa em vez de
                /// sumir de gume no escuro. O fundo pega a luz que entra pela
                /// abertura; as paredes ficam na própria sombra.
                let tom: Color = switch lado {
                case .capa: Color(hex: 0x26262E)
                case .contracapa: Color(hex: 0x1C1C22)
                default: Color(hex: 0x121217)
                }
                contexto.fill(quadrilatero(pontos), with: .color(tom))

                /// ## As peças do interior
                ///
                /// ⚠️ Um estojo aberto **não é liso por dentro**: o keep case tem o
                /// cubo que prende o disco no fundo; o clamshell tem o berço onde o
                /// cassete assenta. Sem elas o interior é «uma textura morta».
                ///
                /// Elas saem de interpolação bilinear dos **mesmos quatro cantos** —
                /// a perspectiva vem de graça da própria face, sem matriz nova.
                if lado == .contracapa { pecaDoInterior(pontos, em: contexto) }
            }

            /// ⚠️ **O lábio da concha** — a meia-lateral que viaja com a tampa. Sem
            /// ela só a capa gira, e o olho lê uma tampa chapada colada numa caixa
            /// maciça. Uma caixa são duas conchas presas na lombada.
            let labio = cantosDaMeiaLateral(medidas, abertura: abertura).map { canto in
                let p = projetado(girado(canto, poseAtual), distancia: distancia)
                return CGPoint(x: p.x + centro.x, y: p.y + centro.y)
            }
            contexto.fill(quadrilatero(labio),
                          with: .color(ehEscuro ? Color(hex: 0x101014) : Color(hex: 0x2A2622)))
        }
        .allowsHitTesting(false)
    }

    /// O cubo do disco, ou o berço do cassete.
    private func pecaDoInterior(_ pontos: [CGPoint], em contexto: GraphicsContext) {
        func ponto(_ u: CGFloat, _ v: CGFloat) -> CGPoint {
            let topo = CGPoint(x: pontos[0].x + (pontos[1].x - pontos[0].x) * u,
                               y: pontos[0].y + (pontos[1].y - pontos[0].y) * u)
            let base = CGPoint(x: pontos[3].x + (pontos[2].x - pontos[3].x) * u,
                               y: pontos[3].y + (pontos[2].y - pontos[3].y) * u)
            return CGPoint(x: topo.x + (base.x - topo.x) * v, y: topo.y + (base.y - topo.y) * v)
        }

        if interiorDeDisco {
            /// O cubo: o anel que prende o disco, no centro do fundo.
            var anel = Path()
            let passos = 28
            for i in 0 ... passos {
                let a = CGFloat(i) / CGFloat(passos) * 2 * .pi
                let p = ponto(0.5 + cos(a) * 0.16, 0.5 + sin(a) * 0.11)
                if i == 0 { anel.move(to: p) } else { anel.addLine(to: p) }
            }
            anel.closeSubpath()
            contexto.stroke(anel, with: .color(.white.opacity(0.14)), lineWidth: 2)
        } else {
            /// O berço: a cavidade retangular onde o cassete assenta.
            var berco = Path()
            berco.move(to: ponto(0.10, 0.14))
            berco.addLine(to: ponto(0.90, 0.14))
            berco.addLine(to: ponto(0.90, 0.86))
            berco.addLine(to: ponto(0.10, 0.86))
            berco.closeSubpath()
            contexto.fill(berco, with: .color(.black.opacity(0.35)))
            contexto.stroke(berco, with: .color(.white.opacity(0.10)), lineWidth: 1.5)
        }
    }

    private func quadrilatero(_ p: [CGPoint]) -> Path {
        var caminho = Path()
        caminho.move(to: p[0])
        for i in 1 ..< p.count { caminho.addLine(to: p[i]) }
        caminho.closeSubpath()
        return caminho
    }

    private var arrasto: some Gesture {
        DragGesture(minimumDistance: 2)
            .onChanged { g in
                inercia?.cancel()
                let base = poseAoPegar ?? poseAtual
                if poseAoPegar == nil { poseAoPegar = base }
                /// ⚠️ **Subtrai** o arrasto. Somando, «parece que a caixa vai pro
                /// lado contrário do dedo» — foi o que o dono viu no Android.
                pose = base.somando(
                    dx: -g.translation.width * Pose.grauPorPixelHorizontal,
                    dy: -g.translation.height * Pose.grauPorPixelVertical,
                    livre: true,
                )
            }
            .onEnded { g in
                poseAoPegar = nil
                /// A velocidade do gesto vira giro que decai — `predictedEnd`
                /// menos o fim real é o quanto o dedo ainda estava andando.
                let sobra = g.predictedEndTranslation.width - g.translation.width
                var velocidade = -sobra * Pose.grauPorPixelHorizontal / 12
                inercia = Task { @MainActor in
                    /// ⚠️ Para quando fica **imperceptível**, e não em zero: uma
                    /// exponencial nunca chega a zero, e o laço rodaria pra sempre
                    /// gastando bateria por um giro que ninguém vê.
                    while abs(velocidade) > 0.05 {
                        pose = pose.somando(dx: velocidade, dy: 0, livre: true)
                        velocidade *= 0.94
                        try? await Task.sleep(for: .milliseconds(16))
                        if Task.isCancelled { return }
                    }
                }
            }
    }
}
